### ISSUE TITLE
Metadata extraction fails after updating to VS15.8

### ISSUE THREAD

**Issue by jskeet** [dotnet/docfx#3225]
A few days ago, all was well - I was using VS15.7.* and docfx 2.32.1. After updating to VS15.8, I received an MSBuild error with docfx (I can post more details if they're useful) so I updated to docfx 23.8.1. That changed the issue to the one given here.

**DocFX Version Used**: 2.38.1

**Template used**: (`default` or `statictoc` or contain custom template)

n/a

**Steps to Reproduce**:

1. Create a new .NET Core project:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
  </PropertyGroup>
</Project>
```

2. Create Class1.cs:

using System;

```csharp
namespace VS158
{
    /// <summary>This is a class</summary>
    public class Class1
    {
    }
}
```

3. Create docfx.json:

```json
{
  "metadata": [
    {
      "src": [ { "files": [ "VS158.csproj" ] } ],
      "dest": "obj/api"
    }
  ]
}
```

4. Run `docfx metadata`

**Expected Behavior**:

Successful metadata extraction.

**Actual Behavior**:

The build effectively fails - it claims it succeeds, but doesn't load any metadata, due to a problem loading System.Collections.Immutable:

(Timestamps removed for clarity)

```text
Info:Config file docfx.json found, start generating metadata...
Info:[ExtractMetadata]Environment variable VSINSTALLDIR is set to C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\, it is used as the inner compiler.
Info:[ExtractMetadata]Loading projects...
Warning:[ExtractMetadata](c:/Users/jon/Test/Projects/Issues/docfx/VS158/VS158.csproj)Workspace failed with: [Failure] Msbuild failed when processing the file 'c:\Users\jon\Test\Projects\Issues\docfx\VS158\VS158.csproj' with message: Could not load file or assembly 'System.Collections.Immutable, Version=1.2.3.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
Warning:[ExtractMetadata]Project 'c:\Users\jon\Test\Projects\Issues\docfx\VS158\VS158.csproj' does not contain any documents.
Info:[ExtractMetadata]Generating metadata for each project...
Warning:[ExtractMetadata]No metadata is generated for VS158.
Info:Completed in 3025.3817 milliseconds
```

**Comment by jskeet** [comment]
I now have a workaround for this, but I don't know how reliable it is:

- Download [System.Collections.Immutable version 1.5.0](https://www.nuget.org/api/v2/package/System.Collections.Immutable/1.5.0)
- Rename the nupkg file to a zip file so it's easy to open
- Copy lib\netstandard1.3\System.Collections.Immutable.dll over the copy in docfx
- Edit docfx.config so that the binding redirect for System.Collections.Immutable uses 1.2.3 instead of 1.2.1

(I don't know why the NuGet package 1.5.0 contains a DLL with a version of 1.2.3, but hey...)

That works with my test program - I'll see if it works on "real" code.

**Comment by HappyTreesDev** [comment]
Any update as to when this would be expected to be fixed? I would rather not have to deal with a work around if possible.

**Comment by lextm** [comment]
@HappyTreesDev docfx (2.38.1) has a dependency on MSBuild 15.7 today, #3231 .

I think a proper fix would come once MSBuild team publishes MSBuild 15.8 on NuGet.org and docfx team upgrades to that version. Unfortunately, history tells us that won't happen soon.

**Comment by herohua** [comment]
Fixed in v2.39.

**Comment by jorgy343** [comment]
If anyone comes across this issue and docfx is complaining about System.Collections.Immutable 1.2.1.0, I found that using version 2.38.1 works. Any version after that will fail to build.

**Comment by AFFogarty** [comment]
@herohua I'm getting this error again in 2.58+.

**Comment by rasmuseeg** [comment]
My solution was to use docfx version 2.57.2 with msbuild included in VS 2022 17.0.0.

**Comment by jeffward01** [comment]
> My solution was to use docfx version 2.57.2 with msbuild included in VS 2022 17.0.0.

I just wanted to chime in and say I am having issues with using the versions of docfx recommended. However, I am using VS 2022 v 17.0.5

Please see Screenshots:

![image](https://user-images.githubusercontent.com/12589359/149610873-668bc1ed-757b-46f4-bf3d-e74a8a750a65.png)

![image](https://user-images.githubusercontent.com/12589359/149610905-97551e21-3106-4285-917b-ec1bdd8b7f5e.png)


This is the target build project:

![image](https://user-images.githubusercontent.com/12589359/149610923-fc392c62-615d-41e8-b046-ea4c316ed19d.png)

My docfx.json:

````json
{
  "metadata": [
    {
      "src": [
        {
          "files": "LibMain/lib/**.csproj",

          "exclude": [
            "**/obj/**",
            "**/bin/**"
          ],
          "src": "../"
        }

      ],
      "dest": "api/Aeonic.Booleans"
    }  
  ],
  "build": {
    "content": [
      {
        "files": [
          "api/**.yml",
          "api/index.md"
        ]
      },
      {
        "files": [
          "articles/**.md",
          "articles/**/toc.yml",
          "toc.yml",
          "*.md"
        ]
      }
    ],
    "resource": [
      {
        "files": [
          "images/**"
        ]
      }
    ],
    "overwrite": [
      {
        "files": [
          "apidoc/**.md"
        ],
        "exclude": [
          "obj/**",
          "_site/**"
        ]
      }
    ],
    "dest": "_site",
    "globalMetadataFiles": [],
    "fileMetadataFiles": [],
    "template": 
    ["default","templates/discordfx"],
    
    "postProcessors": [],
    "markdownEngineName": "markdig",
    "noLangKeyword": false,
    "keepFileLink": false,
    "cleanupCacheHistory": false,
    "disableGitFeatures": false
  }
}

````


Command Output:

````command

 'C:\Dev\Aeonic\AeonicTech\Libraries\Common\Aeonic.Booleans\LibMain\lib\Aeonic.Booleans.csproj' with message: SDK Resolver Failure: "The SDK resolver "Microsoft.DotNet.MSBuildSdkResolver" failed while attempting to resolve the SDK "Microsoft.NET.Sdk". Exception: "System.IO.FileNotFoundException: Could not load file or assembly 'System.Collections.Immutable, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
File name: 'System.Collections.Immutable, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
   at Microsoft.NET.Sdk.WorkloadMSBuildSdkResolver.CachingWorkloadResolver.Resolve(String sdkReferenceName, String dotnetRootPath, String sdkVersion, String userProfileDir)
   at Microsoft.DotNet.MSBuildSdkResolver.DotNetMSBuildSdkResolver.Resolve(SdkReference sdkReference, SdkResolverContext context, SdkResultFactory factory)
   at Microsoft.Build.BackEnd.SdkResolution.SdkResolverService.ResolveSdk(Int32 submissionId, SdkReference sdk, LoggingContext loggingContext, ElementLocation sdkReferenceLocation, String solutionPath, String projectPath, Boolean interactive, Boolean isRunningInVisualStudio)

````
